src/sumo_store_riak.erl

```diff
@@ -0,0 +1,421 @@
+  find_by(DocName, [], Limit, Offset, State).
+
+-spec find_by(sumo:schema_name(), sumo:conditions(), state()) ->
+  sumo_store:result([sumo_internal:doc()], state()).
+find_by(DocName, Conditions, State) ->
+  find_by(DocName, Conditions, 0, 0, State).
+
+-spec find_by(
+  sumo:schema_name(),
+  sumo:conditions(),
+  non_neg_integer(),
+  non_neg_integer(),
+  state()
+) -> sumo_store:result([sumo_internal:doc()], state()).
+find_by(DocName,
+        Conditions,
+        Limit,
+        Offset,
+        #state{conn = Conn, bucket = Bucket, index = Index} = State) ->
+  IdField = sumo_internal:id_field_name(DocName),
+  case lists:keyfind(IdField, 1, Conditions) of
+    {_K, Key} ->
+      case fetch_map(Conn, Bucket, iolist_to_binary(Key)) of
+        {ok, RMap} ->
+          Val = rmap_to_doc(DocName, RMap),
+          {ok, [Val], State};
+        {error, Error} ->
+          {error, Error, State}
+      end;
+    _ ->
+      Query = build_query(Conditions),
+      case search_docs_by(DocName, Conn, Index, Query, Limit, Offset) of
+        {ok, {_, Res}} -> {ok, Res, State};
+        {error, Error} -> {error, Error, State}
+      end
+  end.
+
+-spec find_by(
+  sumo:schema_name(),
+  sumo:conditions(),
+  term(),
+  non_neg_integer(),
+  non_neg_integer(),
+  state()
+) -> sumo_store:result([sumo_internal:doc()], state()).
+find_by(_DocName, _Conditions, _SortFields, _Limit, _Offset, State) ->
+  {error, not_supported, State}.
+
+-spec create_schema(
+  sumo:schema(), state()
+) -> sumo_store:result(state()).
+create_schema(_Schema, State) ->
+  {ok, State}.
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%% Private API.
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+%% @private
+doc_id(Doc) ->
+  DocName = sumo_internal:doc_name(Doc),
+  IdField = sumo_internal:id_field_name(DocName),
+  sumo_internal:get_field(IdField, Doc).
+
+%% @private
+new_doc(Doc, #state{conn = Conn, bucket = Bucket}) ->
+  DocName = sumo_internal:doc_name(Doc),
+  IdField = sumo_internal:id_field_name(DocName),
+  Id = case sumo_internal:get_field(IdField, Doc) of
+         undefined ->
+           case update_map(Conn, Bucket, undefined, doc_to_rmap(Doc)) of
+             {ok, RiakMapId} -> RiakMapId;
+             {error, Error}  -> throw(Error);
+             _               -> throw(unexpected)
+           end;
+         Id0 ->
+           to_bin(Id0)
+       end,
+  {Id, sumo_internal:set_field(IdField, Id, Doc)}.
+
+%% @private
+doc_to_rmap(Doc) ->
+  Fields = sumo_internal:doc_fields(Doc),
+  map_to_rmap(Fields).
+
+%% @private
+%% Support multi-level structures.
+map_to_rmap(Map) ->
+  F = fun({K, V}, Acc) ->
+        case V of
+          V when is_map(V) ->
+            NewV = map_to_rmap(V),
+            riakc_map:update({to_bin(K), map}, fun(_M) -> NewV end, Acc);
+          V when is_list(V) ->
+            case io_lib:printable_list(V) of
```

Comment:
The expression on line 243 and column 13 is nested beyond the maximum level of 3.